### PR TITLE
[2.x] Min PHP version 8.1, Symfony 6.4 / 7.1+

### LINKED ISSUE #458: raise minimum php and symfony versions
for version 2, we should raise the PHP version.

Lets do ^8.1 and only allow Symfony : ^6.4|^7.1. 

Check code for BC things for old symfony versions.

### PR DESCRIPTION
**Pull request by andrew-demb**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | yes
| Deprecations?   | no
| Related tickets | fixes #458 
| Documentation   | -
| License         | MIT

#### What's in this PR?

Change the minimum PHP version to 8.1 and Symfony component versions to 6.4 / 7.1+ as requested in #458 

#### Why?

#458 

#### Example Usage

N/A

#### Checklist

- [x] Updated CHANGELOG.md to describe BC breaks / deprecations | new feature | bugfix
- [ ] Documentation pull request created (if not simply a bugfix)

#### To Do

N/A

### CODE REVIEW

**Comment by andrew-demb** [comment]
> Check code for BC things for old symfony versions.

Didn't find any of them.

**Comment by andrew-demb** [comment]
I need help with this failed CI check

```
PHPUnit 9.6.20 by Sebastian Bergmann and contributors.

Warning:       Your XML configuration validates against a deprecated schema.
Suggestion:    Migrate your XML configuration using "--migrate-configuration"!

Testing 
Call to undefined method Http\HttplugBundle\Tests\Unit\ClientFactory\BuzzFactoryTest::getAnnotations()

Remaining indirect deprecation notices (1)
Script vendor/bin/simple-phpunit handling the test event returned with error code 2
```

I can't find `getAnnotations` call in the vendor that could lead to such error and cannot reproduce this locally

File: src/Collector/Collector.php [hunk 1]
```diff
@@ -21,34 +21,26 @@
...
-    /**
-     * {@inheritdoc}
-     */
+    public function __wakeup(): void
+    {
+        $this->capturedBodyLength = null;
+
+        parent::__wakeup();
+    }
+
```
Comment (ostrolucky) [inline src/Collector/Collector.php:37]: Why was this added? Did fixer do it?

Comment (andrew-demb) [reply src/Collector/Collector.php:37]: Rector inlined property to use constructor property promotion.
But after that, tests that trigger opening profiler result (which unserializes serialized before collector) failed with error "trying to get uninitialized property".

I can ddop using CPP here or leave a logic to initiale a property in wakeup

**Review by ostrolucky — APPROVED** [review]

File: composer.json [hunk 2]
```diff
@@ -57,15 +57,16 @@
...
-        "symfony/browser-kit": "^5.4 || ^6.0 || ^7.0",
-        "symfony/cache": "^5.4 || ^6.0 || ^7.0",
-        "symfony/dom-crawler": "^5.4 || ^6.0 || ^7.0",
-        "symfony/framework-bundle": "^5.4 || ^6.0 || ^7.0",
-        "symfony/http-foundation": "^5.4 || ^6.0 || ^7.0",
-        "symfony/phpunit-bridge": "^7.0.3",
-        "symfony/stopwatch": "^5.4 || ^6.0 || ^7.0",
-        "symfony/twig-bundle": "^5.4 || ^6.0 || ^7.0",
-        "symfony/web-profiler-bundle": "^5.4 || ^6.0 || ^7.0",
+        "phpunit/phpunit": "^9",
```
Comment (dbu) [inline composer.json:61]: when we use the phpunit-bridge, we don't need the phpunit dependency. in 1.x i just ditched the bridge because it ended up with incompatible versions, we should do that in 2.x too (but use a more recent version than 9 if we can)

File: src/Collector/Formatter.php [hunk 3]
```diff
@@ -23,20 +22,8 @@
...
-     */
-    private $formatter;
-
-    /**
-     * @var CurlCommandFormatter
-     */
-    private $curlFormatter;
-
-    public function __construct(MessageFormatter $formatter, MessageFormatter $curlFormatter)
+    public function __construct(private readonly MessageFormatter $formatter, private readonly MessageFormatter $curlFormatter)
```
Comment (dbu) [inline src/Collector/Formatter.php]: with multiple promoted constructor arguments, i find it much more readable to have each property on its own line:

```suggestion
    public function __construct(
        private readonly MessageFormatter $formatter, 
        private readonly MessageFormatter $curlFormatter,
    )
```

File: src/Collector/PluginClientFactory.php [hunk 4]
```diff
@@ -20,26 +20,8 @@
...
-     */
-    private $formatter;
-
-    /**
-     * @var Stopwatch
-     */
-    private $stopwatch;
-
-    public function __construct(Collector $collector, Formatter $formatter, Stopwatch $stopwatch)
+    public function __construct(private readonly Collector $collector, private readonly Formatter $formatter, private readonly Stopwatch $stopwatch)
```
Comment (dbu) [inline src/Collector/PluginClientFactory.php]:
```suggestion
    public function __construct(
        private readonly Collector $collector,
        private readonly Formatter $formatter,
        private readonly Stopwatch $stopwatch,
    )
```

File: src/Collector/Profile.php [hunk 5]
```diff
@@ -13,32 +13,14 @@
  */
 final class Profile
 {
-    /**
-     * @var string
-     */
-    private $plugin;
+    private string $request;
```
Comment (dbu) [inline src/Collector/Profile.php]: this changes the behaviour: if we try to getRequest before a request is set, php will do a fatal error because an unitialized property is accessed.

i think we should initialize with null to keep the rendering of the profile the same as before:

```suggestion
    private ?string $request = null;
```

File: src/Collector/Profile.php [hunk 6]
```diff
@@ -13,32 +13,14 @@
...
 
-    /**
-     * @var string
-     */
-    private $request;
+    private string $response;
```
Comment (dbu) [inline src/Collector/Profile.php]: same here
```suggestion
    private ?string $response = null;
```

File: src/Collector/ProfileClient.php [hunk 7]
```diff
@@ -26,56 +26,32 @@ class ProfileClient implements ClientInterface, HttpAsyncClient
...
+    private array $eventNames = [];
 
     private const STOPWATCH_CATEGORY = 'httplug';
 
     /**
      * @param ClientInterface|HttpAsyncClient $client The client to profile. Client must implement HttpClient or
      *                                                HttpAsyncClient interface.
      */
-    public function __construct($client, Collector $collector, Formatter $formatter, Stopwatch $stopwatch)
+    public function __construct($client, private readonly Collector $collector, private readonly Formatter $formatter, private readonly Stopwatch $stopwatch)
```
Comment (dbu) [inline src/Collector/ProfileClient.php]: separate lines. and $client can be typehinted [since php 8.0](https://3v4l.org/Wnp4m)
```suggestion
    public function __construct(
        ClientInterface|HttpAsyncClient $client, 
        private readonly Collector $collector, 
        private readonly Formatter $formatter, 
        private readonly Stopwatch $stopwatch,
    )
```

File: src/Collector/ProfileClientFactory.php [hunk 8]
```diff
@@ -24,38 +24,17 @@ class ProfileClientFactory implements ClientFactory
...
-    /**
-     * @var Stopwatch
-     */
-    private $stopwatch;
-
     /**
      * @param ClientFactory|callable $factory
      */
-    public function __construct($factory, Collector $collector, Formatter $formatter, Stopwatch $stopwatch)
+    public function __construct($factory, private readonly Collector $collector, private readonly Formatter $formatter, private readonly Stopwatch $stopwatch)
```
Comment (dbu) [inline src/Collector/ProfileClientFactory.php]:
```suggestion
    public function __construct(
        ClientFactory|callable $factory,
        private readonly Collector $collector,
        private readonly Formatter $formatter,
        private readonly Stopwatch $stopwatch,
    )
```

File: src/Collector/ProfileClientFactory.php [hunk 9]
```diff
@@ -24,38 +24,17 @@ class ProfileClientFactory implements ClientFactory
...
     {
         if (!$factory instanceof ClientFactory && !is_callable($factory)) {
             throw new \RuntimeException(sprintf('First argument to ProfileClientFactory::__construct must be a "%s" or a callable.', ClientFactory::class));
```
Comment (dbu) [inline src/Collector/ProfileClientFactory.php]: obsolete with the parameter type declaration
```suggestion
```

File: src/Collector/ProfilePlugin.php [hunk 10]
```diff
@@ -21,31 +20,13 @@ class ProfilePlugin implements Plugin
...
-     */
-    private $collector;
-
-    /**
-     * @var Formatter
-     */
-    private $formatter;
-
-    public function __construct(Plugin $plugin, Collector $collector, Formatter $formatter)
+    public function __construct(private Plugin $plugin, private Collector $collector, private Formatter $formatter)
```
Comment (dbu) [inline src/Collector/ProfilePlugin.php]:
```suggestion
    public function __construct(
        private Plugin $plugin,
        private Collector $collector,
        private Formatter $formatter,
    )
```

File: src/Collector/Stack.php [hunk 11]
```diff
@@ -13,96 +13,44 @@
  */
 final class Stack
 {
-    /**
-     * @var string
-     */
-    private $client;
+    private string $client;
```
Comment (dbu) [inline src/Collector/Stack.php]: constructor argument promotion
```suggestion
```

File: src/Collector/Stack.php [hunk 12]
```diff
@@ -13,96 +13,44 @@
...
      * @var Profile[]
      */
-    private $profiles = [];
+    private array $profiles = [];
 
-    /**
-     * @var string
-     */
-    private $request;
+    private string $request;
```
Comment (dbu) [inline src/Collector/Stack.php]: constructor argument promotion
```suggestion
```

File: src/Collector/Stack.php [hunk 13]
```diff
@@ -13,96 +13,44 @@
...
-     */
-    private $curlCommand;
+    private ?string $curlCommand = null;
 
-    /**
-     * @param string $client
-     * @param string $request
-     */
-    public function __construct($client, $request)
+    public function __construct(string $client, string $request)
```
Comment (dbu) [inline src/Collector/Stack.php]:
```suggestion
    public function __construct(
        private string $client,
        private string $request,
    )
```

File: src/Collector/Stack.php [hunk 14]
```diff
@@ -13,96 +13,44 @@
...
     {
         $this->client = $client;
```
Comment (dbu) [inline src/Collector/Stack.php]:
```suggestion
```

File: src/Collector/Stack.php [hunk 15]
```diff
@@ -13,96 +13,44 @@
...
         $this->request = $request;
```
Comment (dbu) [inline src/Collector/Stack.php]:
```suggestion
```

File: src/Collector/Twig/HttpMessageMarkupExtension.php [hunk 16]
```diff
@@ -16,32 +16,24 @@
...
+    private ClonerInterface $cloner;
 
-    /**
-     * @var HtmlDumper
-     */
-    private $dumper;
+    private DataDumperInterface $dumper;
 
-    public function __construct(?ClonerInterface $cloner = null, ?DataDumperInterface $dumper = null)
+    public function __construct(ClonerInterface $cloner = null, DataDumperInterface $dumper = null)
```
Comment (dbu) [inline src/Collector/Twig/HttpMessageMarkupExtension.php]: afaik php 8.3 requires nullable to always have `?`
```suggestion
    public function __construct(?ClonerInterface $cloner = null, ?DataDumperInterface $dumper = null),
```

**Review by dbu — COMMENTED** [review]
yay for modern php, looks very nice! thanks for the work!

about the failing lowest version test: we should try with `"matthiasnoback/symfony-dependency-injection-test": "^5.1",` and if that is not enough add `matthiasnoback/symfony-config-test` with ^5.2.

i have some minor feedback. do you have time to update these things?

File: composer.json (hunk 2 above)
Comment (dbu) [reply composer.json:61]: for this pull request, i suggest removing the line again. i will then cleanup the 2.x branch by merging 1.x to 2.x

Comment (andrew-demb) [reply composer.json:61]: I added an explanation of why it was added in commit https://github.com/php-http/HttplugBundle/pull/463/commits/0725be77660379b018bbbe31deec26084be83a7a

Without this line - composer installs PHPUnit 10 which is incompatible with the `phpunit-bridge`.

So I can remove it only if I will stick `matthiasnoback/symfony-dependency-injection-test` for a version that is compatible only with PHPUnit 9

File: src/Collector/Twig/HttpMessageMarkupExtension.php (hunk 16 above)
Comment (andrew-demb) [reply src/Collector/Twig/HttpMessageMarkupExtension.php]: ++

I changed php-cs-fixer config to accomplish that

File: src/Collector/Stack.php (hunk 15 above)
Comment (andrew-demb) [reply src/Collector/Stack.php]: ++

File: src/Collector/Stack.php (hunk 14 above)
Comment (andrew-demb) [reply src/Collector/Stack.php]: ++

File: src/Collector/Stack.php (hunk 13 above)
Comment (andrew-demb) [reply src/Collector/Stack.php]: ++

File: src/Collector/Stack.php (hunk 12 above)
Comment (andrew-demb) [reply src/Collector/Stack.php]: ++

File: src/Collector/Stack.php (hunk 11 above)
Comment (andrew-demb) [reply src/Collector/Stack.php]: ++

File: src/Collector/ProfilePlugin.php (hunk 10 above)
Comment (andrew-demb) [reply src/Collector/ProfilePlugin.php]: ++

File: src/Collector/ProfileClientFactory.php (hunk 9 above)
Comment (andrew-demb) [reply src/Collector/ProfileClientFactory.php]: ++

File: src/Collector/ProfileClientFactory.php (hunk 8 above)
Comment (andrew-demb) [reply src/Collector/ProfileClientFactory.php]: ++

File: src/Collector/ProfileClient.php (hunk 7 above)
Comment (andrew-demb) [reply src/Collector/ProfileClient.php]: ++

File: src/Collector/Profile.php (hunk 6 above)
Comment (andrew-demb) [reply src/Collector/Profile.php]: ++

File: src/Collector/Profile.php (hunk 5 above)
Comment (andrew-demb) [reply src/Collector/Profile.php]: ++

File: src/Collector/PluginClientFactory.php (hunk 4 above)
Comment (andrew-demb) [reply src/Collector/PluginClientFactory.php]: ++

File: src/Collector/Formatter.php (hunk 3 above)
Comment (andrew-demb) [reply src/Collector/Formatter.php]: ++

File: composer.json (hunk 2 above)
Comment (dbu) [reply composer.json:61]: ah, i see.

i will try to clean this up in a separate PR (getting rid of phpunit-bridge and upgrading phpunit if possible)

File: src/Collector/ProfilePlugin.php (hunk 10 above)
Comment (dbu) [reply src/Collector/ProfilePlugin.php]: you did not actually commit this one ;-)

**Comment by dbu** [comment]
i think we need to add `"matthiasnoback/symfony-config-test": ^5.2"` to require-dev to fix the lowest build. and change the symfony-dependency-injection-test to only allow ^5.1

**Comment by dbu** [comment]
and RE BC hacks: there is one in PluginClientFactoryListener, search the code for `Kernel::MAJOR_VERSION` to see if there are maybe other places.

File: src/Collector/ProfilePlugin.php (hunk 10 above)
Comment (andrew-demb) [reply src/Collector/ProfilePlugin.php]: Sorry, replaced with regexp mentioned "readonly"

**Comment by andrew-demb** [comment]
> and RE BC hacks: there is one in PluginClientFactoryListener, search the code for Kernel::MAJOR_VERSION to see if there are maybe other places.

I'm sorry, I don't seem to get it.

I failed to see any compatibility layer with old Symfony versions in `PluginClientFactoryListener`
And can't find any usage of the `Kernel::VERSION*` constants usage in the codebase here.

I checked before any usage of deprecated methods from Symfony, `method_exists` of `instanceof` for the old versions of Symfony.

**Comment by dbu** [comment]
> BC hacks

oh, so sorry, that has already been cleaned up in the 2.x branch, i must have looked at 1.x :facepalm: 


thanks a lot for this work, will now merge it :+1:
